### DISCUSSION TITLE
Snippet in autocomplete

### DISCUSSION THREAD

**Discussion by mr-scrpt** [VonHeikemen/lsp-zero.nvim#282] [Q&A]
Hi all. Can you please tell me how i can get snippets for react in autocomplete. Before I was using coc and it had a coc-snippet module and it gave me snippets for react, but now I switched to lsp zero and am trying to figure everything out - here is my config
```

local lsp = require("lsp-zero")
lsp.preset("recommended")

lsp.ensure_installed({
  'tsserver',
})

-- Fix Undefined global 'vim'
lsp.configure('sumneko_lua', {
    settings = {
        Lua = {
            diagnostics = {
                globals = { 'vim' }
            }
        }
    }
})

lsp.setup_nvim_cmp({
  mapping = cmp_mappings,
  sources = {
    {name = 'emmet_ls' },
    {name = 'nvim_lsp' },
    {name = 'copilot'},
  },
  experimental = {
    ghost_text = true,
  },

  completete = false
})

lsp.set_preferences({
    suggest_lsp_servers = false,
    set_lsp_keymaps = false,
    sign_icons = {
        error = '☢️',
        warn = '⚠️',
        hint = 'H',
        info = 'I'
    }
})

lsp.on_attach(function(_, bufnr)
  local opts = {buffer = bufnr, remap = false}

  vim.keymap.set("n", "gt", function() vim.lsp.buf.definition() end, opts)
  vim.keymap.set("n", "gk", function() vim.lsp.buf.hover() end, opts)
  vim.keymap.set("n", "<leader>vws", function() vim.lsp.buf.workspace_symbol() end, opts)
  vim.keymap.set("n", "<leader>vd", function() vim.diagnostic.open_float() end, opts)
  vim.keymap.set("n", "[d", function() vim.diagnostic.goto_next() end, opts)
  vim.keymap.set("n", "]d", function() vim.diagnostic.goto_prev() end, opts)
  vim.keymap.set("n", "<leader>vca", function() vim.lsp.buf.code_action() end, opts)
  vim.keymap.set("n", "<leader>vrr", function() vim.lsp.buf.references() end, opts)
  vim.keymap.set("n", "<leader>vrn", function() vim.lsp.buf.rename() end, opts)
  vim.keymap.set("i", "<C-h>", function() vim.lsp.buf.signature_help() end, opts)
end)

lsp.setup()

vim.diagnostic.config({
    virtual_text = true
})

```
I tried everything as described here, but it didn't work.
[https://github.com/VonHeikemen/lsp-zero.nvim/blob/v2.x/doc/md/autocomplete.md#add-an-external-collection-of-snippets](https://github.com/VonHeikemen/lsp-zero.nvim/blob/v2.x/doc/md/autocomplete.md#add-an-external-collection-of-snippets)

**Comment by VonHeikemen** [comment]
What version of lsp-zero are you using? you can check by executing the command `:help lsp-zero-version`. Note that the current version is `v2.x`.

**Reply by mr-scrpt** [reply]
Thank you for your reply. Was the first branch, now went to the second and it looks like my emmet doesn't work. Here is the code for lsp zero
```
local lsp = require('lsp-zero').preset({})

lsp.on_attach(function(client, bufnr)
  lsp.default_keymaps({buffer = bufnr})
end)

-- When you don't have mason.nvim installed
-- You'll need to list the servers installed in your system
lsp.setup_servers({'tsserver', 'eslint'})

-- (Optional) Configure lua language server for neovim
require('lspconfig').lua_ls.setup(lsp.nvim_lua_ls())

lsp.setup()

local cmp = require('cmp')
local cmp_action = require('lsp-zero').cmp_action()

require('luasnip.loaders.from_vscode').lazy_load()

cmp.setup({
  sources = {
    -- {name = 'path'},
        {name = 'emmet_ls'},
        {name = 'nvim_lsp'},
        {name = 'luasnip'},
        {name = 'copilot'},
    -- {name = 'buffer', keyword_length = 3},
  },
  mapping = {
    ['<CR>'] = cmp.mapping.confirm({select = true}),
    ['<C-f>'] = cmp_action.luasnip_jump_forward(),
    ['<C-b>'] = cmp_action.luasnip_jump_backward(),
  }
})

```
For example, I open an html file and try to write "html5" or "!" and this is what I get
<img width="297" alt="image" src="https://github.com/VonHeikemen/lsp-zero.nvim/assets/48452376/ef1f3b27-0c78-453b-b641-7b96739767c9">
If I confirm the autocomplete, this is what I get
<img width="398" alt="image" src="https://github.com/VonHeikemen/lsp-zero.nvim/assets/48452376/cf48b68b-cea0-4dec-8261-adccd57477f2">
But on the first branch it unfolded into an blank html template, it seems to me that "emmet_ls" has stopped working
Also, if I write "rfc" (react function component) in the .tsx file, I will not get a snippet for react as I did with coc-snippets. Maybe I need to install something else?

**Reply by VonHeikemen** [reply]
Are you using `mason.nvim`? If the answer is no, then you need to add it to the list of servers in `.setup_servers()`.

```lua
lsp.setup_servers({'tsserver', 'eslint', 'emmet_ls'})
```

**Reply by mr-scrpt** [reply]
Yes, I use mason, and I also have emmet_ls installed.
<img width="799" alt="image" src="https://github.com/VonHeikemen/lsp-zero.nvim/assets/48452376/47312275-e3cc-4a42-a8da-714eaaf97087">

**Reply by VonHeikemen** [reply]
Ok. You might want to delete this:

```lua
-- When you don't have mason.nvim installed
-- You'll need to list the servers installed in your system
lsp.setup_servers({'tsserver', 'eslint'})
```

Is not doing anything useful.

For emmet lsp, try to call it manually with `lspconfig`.

```lua
require('lspconfig').emmet_ls.setup({})
```

Another question, in your autocomplete menu do you see any items with the label `[luasnip]` (in any language)?

**Reply by mr-scrpt** [reply]
Hi, I have made progress with emmet. There are two in the mason list and I can just include them there, I was confused that the autocomplete list doesn't say emmet, it says Snippet, for example if I have emmet-ls enabled, if I write .test in the html file I see ".test~ [LSP] Snippet" listed - and I can expand that into `<div class="test"></div>` - this is what emmet should do, also the mason list has emmet-langue-server, if I enable it, add this to the config
[https://github.com/olrtg/emmet-language-server#neovim](https://github.com/olrtg/emmet-language-server#neovim)
After this setting in the html file write .test then I see in the list of autocomplete ".test~ [LSP] Text" - again no Emmet inscription, but it says TEXT - but I made sure that I get these cues from the LSP which I select in the mason list, but for some reason no autocomplete on the symbol "!" - Emmet should do autocomplete on an empty html template - but it does not happen.
**About luasnip**
Disabled everything except luasnip
```
sources = {
       -- {name = 'path'},
        -- {name = 'emmet_ls', keyword_length = 1 },
        -- {name = 'emmet-language-server', keyword_length = 4},
        -- {name = 'nvim_lsp', keyword_length = 1 },
        {name = 'luasnip', keyword_length = 1 },
        -- {name = 'copilot', keyword_length = 1 },
        -- {name = 'buffer',  keyword_length = 1 },
  },
```

I install it as described in the manual
```
 {
        "L3MON4D3/LuaSnip",
      -- follow latest release.
        tag = "v1.*", -- Replace <CurrentMajor> by the latest released major (first number of latest release)
        -- install jsregexp (optional!:).
        run = "make install_jsregexp"
 },
```
But nothing appears in the autocomplete list. Maybe I need to install something else?

**Comment by mr-scrpt** [comment]
Luasnip was able to start, I reinstalled the plugins, these two and removed the parameters to install luasnip
```
      -- Snippets
      { 'L3MON4D3/LuaSnip' },
      { 'rafamadriz/friendly-snippets' },

      -- Autocompletion
      { 'hrsh7th/nvim-cmp' },
      { 'hrsh7th/cmp-buffer' },
      { 'hrsh7th/cmp-path' },
      {'saadparwaiz1/cmp_luasnip'},
      { 'hrsh7th/cmp-nvim-lsp' `},`
      { 'hrsh7th/cmp-nvim-lua' },
```
Also here is my final config for lsp
```
local lsp = require('lsp-zero').preset({})

lsp.set_preferences({
    suggest_lsp_servers = false,
    set_lsp_keymaps = false,
    sign_icons = {
        error = '☢️',
        warn = '⚠️',
        hint = 'H',
        info = 'I'
    }
})

lsp.on_attach(function(client, bufnr)
  lsp.default_keymaps({buffer = bufnr})
  local opts = {buffer = bufnr, remap = false}
  vim.keymap.set("n", "gt", function() vim.lsp.buf.definition() end, opts)
  vim.keymap.set("n", "gk", function() vim.lsp.buf.hover() end, opts)
  vim.keymap.set("n", "<leader>vws", function() vim.lsp.buf.workspace_symbol() end, opts)
  vim.keymap.set("n", "<leader>vd", function() vim.diagnostic.open_float() end, opts)
  vim.keymap.set("n", "[d", function() vim.diagnostic.goto_next() end, opts)
  vim.keymap.set("n", "]d", function() vim.diagnostic.goto_prev() end, opts)
  vim.keymap.set("n", "<leader>vca", function() vim.lsp.buf.code_action() end, opts)
  vim.keymap.set("n", "<leader>vrr", function() vim.lsp.buf.references() end, opts)
  vim.keymap.set("n", "<leader>vrn", function() vim.lsp.buf.rename() end, opts)
  vim.keymap.set("i", "<C-h>", function() vim.lsp.buf.signature_help() end, opts)
end)

local lspconfig = require('lspconfig')

lspconfig.lua_ls.setup(lsp.nvim_lua_ls())

lsp.setup()


local cmp = require('cmp')
local cmp_action = require('lsp-zero').cmp_action()
local luasnip = require("luasnip")
-- require("luasnip.loaders.from_lua").load({ paths = "~/.config/nvim/snippet/" })
require("luasnip.loaders.from_vscode").lazy_load()


--   פּ ﯟ   some other good icons
local kind_icons = {
	Text = "",
	Method = "m",
	Function = "",
	Constructor = "",
	Field = "",
	Variable = "",
	Class = "",
	Interface = "",
	Module = "",
	Property = "",
	Unit = "",
	Value = "",
	Enum = "",
	Keyword = "",
	Snippet = "",
	Color = "",
	File = "",
	Reference = "",
	Folder = "",
	EnumMember = "",
	Constant = "",
	Struct = "",
	Event = "",
	Operator = "",
	TypeParameter = "",
}


cmp.setup({
  snippet = {
		expand = function(args)
			luasnip.lsp_expand(args.body) -- For `luasnip` users.
		end,
	},

  sources = {
    {name = "luasnip"},
    {name = 'path'},
    {name = 'emmet_ls', keyword_length = 1 },
        -- {name = 'emmet-language-server', keyword_length = 4},
    {name = 'nvim_lsp', keyword_length = 1 },
    {name = 'copilot', keyword_length = 1 },
    {name = 'buffer',  keyword_length = 1 },
  },
  formatting = {
		fields = { "kind", "abbr", "menu" },
		format = function(entry, vim_item)
			-- Kind icons
			vim_item.kind = string.format("%s", kind_icons[vim_item.kind])
			-- vim_item.kind = string.format('%s %s', kind_icons[vim_item.kind], vim_item.kind) -- This concatonates the icons with the name of the item kind
			vim_item.menu = ({
				copilot = "[Copilot]",
				luasnip = "LuaSnip",
				nvim_lua = "[NVim Lua]",
				nvim_lsp = "[LSP]",
				buffer = "[Buffer]",
				path = "[Path]",
			})[entry.source.name]
			return vim_item
		end,
	},
  mapping = {
    ['<CR>'] = cmp.mapping.confirm({select = true}),
    ['<C-f>'] = cmp_action.luasnip_jump_forward(),
    ['<C-b>'] = cmp_action.luasnip_jump_backward(),
  },
  experimental = {
    ghost_text = true,
  },

  completete = false

})
vim.diagnostic.config({
    virtual_text = true
})
```
But Emmet still does not expand the "!"

**Reply by VonHeikemen** [reply]
I think there is a misunderstanding here. The option `sources` in nvim-cmp does not control all LSP completions. Each source in nvim-cmp is a Neovim plugin. For example `{name = 'path'}` actives this plugin [hrsh7th/cmp-path](https://github.com/hrsh7th/cmp-path). So `{name = 'emmet_ls'}` doesn't have any effect, that is not a separate plugin. All LSP completions are controlled by a single source: `{name = 'nvim_lsp'}`, which activates this plugin [hrsh7th/cmp-nvim-lsp](https://github.com/hrsh7th/cmp-nvim-lsp). And to be clear, `{name = 'luasnip'}` actives this plugin [saadparwaiz1/cmp_luasnip](https://github.com/saadparwaiz1/cmp_luasnip).

Try adding a mapping to trigger the completion menu manually.

```lua
    -- Ctrl+Space to trigger completion menu
    ['<C-Space>'] = cmp.mapping.complete(),
```

Add the `!` in your test html, press the mapping. If emmet doesn't suggest anything that's an issue with the LSP server.
